### PR TITLE
Cbrune/bfd 550 - API for enrollment data for year-month data

### PR DESCRIPTION
**Pull request by cbrunefearless**
### Change Details
Added a function that will route based on parameters from the api call.  This will search beneficiaries by yearmonth and part d contract number. This PR is to keep track of enrollment data by month and year.

### Acceptance Validation
Make sure the right data is being passed back from the api call.  Make sure you can pass by Part D and year-month

### Feedback Requested
Review unit test, run locally calling the api and all.  

### External References

- [BFD-550](https://jira.cms.gov/browse/BFD-550)

### Security Implications

No
- [ ] new software dependencies
No

No
- [ ] altered security controls
No

No
- [ ] new data stored or transmitted
No

No
- [ ] security checklist is completed for this change
Yes

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

**Comment by jzulim** [comment]
Thank you for this excellent work @cbrunefearless - I've gone ahead and tagged AB2D. This PR has huge implications into who and what data we release and will benefit from as many reviews and test / functionality scrutiny that we can throw at it.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 1]
```diff
@@ -855,4 +879,239 @@ public static void checkPageSize(LinkBuilder paging) {
...
+        "select distinct b from Beneficiary b "
+            + joinsClause
+            + "where b.beneficiaryId in :ids "
+            + "order by b.beneficiaryId asc";
+    return entityManager.createQuery(query, Beneficiary.class).setParameter("ids", ids);
+  }
+
+  private String partDFieldByMonth(CcwCodebookVariable month) {
+    if (month == CcwCodebookVariable.PTDCNTRCT01) return "01";
+    if (month == CcwCodebookVariable.PTDCNTRCT02) return "02";
```
Comment (cbrunefearless) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:1001]: add static hashmaps

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java [hunk 2]
```diff
@@ -855,4 +879,239 @@ public static void checkPageSize(LinkBuilder paging) {
...
+      TokenParam coverageId,
+      TokenParam referenceYear,
+      String cursor,
+      RequestDetails requestDetails) {
+    checkCoverageId(coverageId);
+    List<String> includeIdentifiersValues = returnIncludeIdentifiersValues(requestDetails);
+    PatientLinkBuilder paging = new PatientLinkBuilder(requestDetails.getCompleteUrl());
+    checkPageSize(paging);
+
+    Operation operation = new Operation(Operation.Endpoint.V1_PATIENT);
```
Comment (cbrunefearless) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4PatientResourceProvider.java:1028]: v2

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 3]
```diff
@@ -330,6 +413,23 @@ private String partDFieldFor(CcwCodebookVariable month) {
...
   }
 
+  private String partDFieldByMonth(CcwCodebookVariable month) {
+    if (month == CcwCodebookVariable.PTDCNTRCT01) return "01";
+    if (month == CcwCodebookVariable.PTDCNTRCT02) return "02";
+    if (month == CcwCodebookVariable.PTDCNTRCT03) return "03";
+    if (month == CcwCodebookVariable.PTDCNTRCT04) return "04";
+    if (month == CcwCodebookVariable.PTDCNTRCT05) return "05";
+    if (month == CcwCodebookVariable.PTDCNTRCT06) return "06";
+    if (month == CcwCodebookVariable.PTDCNTRCT07) return "07";
```
Comment (cbrunefearless) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:423]: static hashmap

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageTransformer.java [hunk 4]
```diff
@@ -929,6 +929,118 @@ private static Coverage transformPartD(MetricRegistry metricRegistry, Beneficiar
...
+              String yearMonth =
+                  String.format(
+                      "%s-%s",
+                      String.valueOf(beneMonthly.getYearMonth().getYear()), String.valueOf(month));
+
+              switch (month) {
+                case 1:
+                  coverage.addExtension(
+                      TransformerUtils.createExtensionCoding(
+                          coverage,
```
Comment (cbrunefearless) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/CoverageTransformer.java:948]: hashmap

**Review by sb-wnyffenegger — COMMENTED** [review]
Why not use a base class for both the stu3 and R4 patient providers since so many methods are duplicated between them?

Also there is a bug on line 823 of the R4PatientResourceProvider where String comparison is being done using `==`. It may also exist in the original PatientResourceProvider.

**Comment by cbrunefearless** [comment]
> Why not use a base class for both the stu3 and R4 patient providers since so many methods are duplicated between them?
> 
> Also there is a bug on line 823 of the R4PatientResourceProvider where String comparison is being done using `==`. It may also exist in the original PatientResourceProvider.

changed to strings.isnullorempty

**Review by sb-wnyffenegger — APPROVED** [review]

**Review by jzulim — APPROVED** [review]
Looks good to me. Another fine piece of work. Simple (in a good way, i.e. method routing), naming looks good, comments sufficient (I do think we might want to use more comments in general, something to mull over) and generally great job!

**Review by sb-wnyffenegger — APPROVED** [review]

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProviderIT.java [hunk 5]
```diff
@@ -1892,6 +1892,131 @@ public void searchWithLastUpdated() {
...
+                        "2010"))
+            .returnBundle(Bundle.class)
+            .execute();
+
+    Assert.assertNotNull(searchResults);
+    Assert.assertEquals(0, searchResults.getEntry().size());
+  }
+
+  @Test
+  public void searchForPatientByPartDContractNumWithAInvalidContract() {
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProviderIT.java:1960]: Should this be: searchForPatientByPartDContractNumAndYearithAInvalidContract

File: apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProviderIT.java [hunk 6]
```diff
@@ -1892,6 +1892,131 @@ public void searchWithLastUpdated() {
     testLastUpdatedUrls(fhirClient, beneficiary.getBeneficiaryId(), emptyUrls, 0);
   }
 
+  @Test
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/test/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProviderIT.java:1895]: We should consider adding comments to step the developer through these tests.

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 7]
```diff
@@ -257,8 +291,7 @@ public Bundle searchByLogicalId(
     return bundle;
   }
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:293]: Should this include @Search ?

File: apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java [hunk 8]
```diff
@@ -310,6 +343,57 @@ public Bundle searchByCoverageContract(
...
 
+  public Bundle searchByCoverageContractAndYearMonth(
```
Comment (jzulim) [inline apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java:346]: Should this include @Search ?

**Review by jzulim — APPROVED** [review]
Let it rip!

**Review by sharshar-sb — APPROVED** [review]

**Review by keithdadkins — APPROVED** [review]

**Pull request merged**